gadopt/energy_solver.py

```diff
@@ -53,143 +51,335 @@
     "pc_type": "lu",
     "pc_factor_mat_solver_type": "mumps",
 }
-"""Default direct solver parameters for solution of energy equation. Configured to use LU factorisation,
-   using the MUMPS library. G-ADOPT defaults to direct solvers in 2-D.
+"""Default direct solver parameters for solution of energy equation.
+
+Configured to use LU factorisation performed via the MUMPS library.
+
+Note:
+  G-ADOPT defaults to direct solvers in 2-D.
 """
 
 
-class EnergySolver:
-    """Timestepper and solver for the energy equation. The temperature, T, is updated in place.
+class MetaPostInit(abc.ABCMeta):
```

Comment:
I am happy to learn more about this but I still don't see why it's really necessary... 

A few commits ago you didnt have them in and it seemed a fairly simple PR change...

I was reading here towards the end 
_'''Now the big question. Why would you use some obscure error-prone feature?

Well, usually you don't:

    Metaclasses are deeper magic that 99% of users should never worry about it. If you wonder whether you need them, you don't (the people who actually need them know with certainty that they need them, and don't need an explanation about why).

Python Guru Tim Peters'''_
https://stackoverflow.com/questions/100003/what-are-metaclasses-in-python

Reply:
I do not think I am wondering whether I need the metaclass here. :)

In `master`, we have a lengthy `__init__` method. This is not optimal when trying to implement a class inheritance logic. So, I started by creating some methods, such as `set_boundary_conditions`, to tidy the `__init__` method. I wanted these methods to lie in an abstract class from which to define subclasses. I require the class to be abstract because subclasses must implement their own `set_equation` method; thus, it makes sense to have the latter as an `abstractmethod`. With the inheritance logic, I realised that the newly implemented methods are quite general and should not require any method overwriting. However, that same inheritance logic implies that it is unlikely one bypasses the overwriting of `__init__` when implementing a new solver. Now the problem starts to shape. The abstract base class `__init__` eventually calls all the methods I defined. You could move these calls outside of `__init__`, but I do not think we want a user to have to call another method after calling the solver's constructor. Given the nature of the inheritance pattern, I want the subclass to call its parent's `__init__` method using `super` but also bring in additional functionalities. That is, I believe, where a problem can occur. What if new functionalities must be defined before the methods are called but after the remainder of the parent's `__init__` code? In other words, what if, whether you called `super` first or last, you still had a problem? The solution I propose using a metaclass addresses this issue. It separates the methods from the rest of the `__init__` code in the parent class and always makes them execute after the subclass's `__init__`. When subclassing the abstract base class, a developer can clearly separate specific initialisation code from generic method calls.

To be honest, this issue arose when I was working on the `stokes_integrators` module in `update_approx` following changes from the viscoelastic branches. However, I found the solution quite neat, so I thought it would be desirable to also have it in the `energy_solver` module.

I do not see why this approach should be considered error-prone. I see it as a feature to ease further solver development.

Reply:
I agree with @w3168. Please think about the long-term maintainability of any changes you're going to commit. If this were your own project, it would be absolutely fine to write code that pushes the boundaries of your knowledge in an attempt to understand the language better. This is a community project however, and its going to be worked on by people with a range of scientific and programming abilities. Not everyone who wants to develop on G-ADOPT for their own research will have your ability or your motivation to explore Python. 'Get good' isn't the right answer to that, as that kind of attitude could very much hurt adoption of G-ADOPT in the long term. After all, G-ADOPT is a thin wrapper to Firedrake, if building on it is too hard, I'm not going to bother and just write my own Firedrake wrapper. This is why larger projects have standards documents that spell out what language features can and can't be used. That kind of thing is way overkill for G-ADOPT, but I hope you see my point. Apologies if I'm stepping out of my lane here, I've had this battle a few times.

On to the actual code, I've not come across metaclasses before, but what you're doing here seems to attempt to emulate the init behaviour of the `dataclass` decorator. Could you just use `@dataclass` and guarantee the order of `__init__` calls and the other methods that way? This looks like a more appropriate use of that decorator than in e.g. `equations.py` where half the stuff set by the default `__init__` method is overwritten in `__post_init__`.

Reply:
```quote
Data classes are just regular classes that are geared towards storing state, rather than containing a lot of logic. Every time you create a class that mostly consists of attributes, you make a data class.
```
On the `dataclass` aspect, I try to follow this statement from [Martijn Pieters](https://stackoverflow.com/a/47955313/10640534). From that perspective, `Equation` seems like an obvious candidate to me but not any G-ADOPT solver. I would add that nothing is overwritten in `Equation`'s `__post_init__`: `InitVar` is part of the language specification, and I use it to replicate the behaviour that was previously in `master`. It makes it clear which arguments are stored as-is and which ones imply some post-processing.

Regarding maintainability, I fail to see the issue. Indeed, metaclasses are not your common Python object, but it is not as if the one I implemented is achieving anything mysterious or unreasonable. The docstring is clear and there is only one added call. As I wrote elsewhere, this change is geared towards making the implementation of solvers easier. Frankly, look at the `EnergySolver` code. Would you argue there is anything complicated or unreasonable in it? From my perspective, it is extremely straightforward, such that anyone who would attempt to add a different solver could easily be inspired by that one and achieve what they want.

Reply:
OK, you've convinced me that `@dataclass` isn't the right construct for this. But lets look at how `@dataclass` implements the `__post_init__` call then. Surely the implementation in the language itself will be the preferred option? [First, it tests if `__post_init__` exists](https://github.com/python/cpython/blob/34653bba644aa5481613f398153757d7357e39ea/Lib/dataclasses.py#L1076), then follows on to `_init_fn` which is dynamically building an `__init__` method for the class in a `_FuncBuilder` instance, [notably adding the `__post_init__` call last](https://github.com/python/cpython/blob/34653bba644aa5481613f398153757d7357e39ea/Lib/dataclasses.py#L649). Once that's done it dynamically creates `__repr__`, `__eq__` and ordering methods, then calls `add_fns_to_class`. Finally, [the whole thing is run with `exec`](https://github.com/python/cpython/blob/34653bba644aa5481613f398153757d7357e39ea/Lib/dataclasses.py#L502).

Obviously I would not advocate for this entire workflow in this case. `@dataclass` needs far more flexibility than we do, and we all know about `exec`. What I note, however, is that there are no metaclasses defined here. It simply creates an `__init__` that runs `__post_init__` last. If Python itself does not use a metaclass to implement this behaviour, we shouldn't either.

Reply:
Very interesting, thank you. How would that work with the subclassing idea though? My whole point is to separate `__post_init__` from `__init__` such that `__post_init__` would always execute after all `__init__` have run (involving `super` in the process).

Reply:
Custom decorator? https://gist.github.com/dsroberts/cbd9ed5b9ac607bf4ddbab171102fdf1

Reply:
You could make the whole thing a little more user friendly by dropping `__post_init__` and naming the function after what its actually doing. So rename `__post_init__` to `solver_initialise` or something and call the decorator `@solver`. Documentation can explain why we have a decorator on those classes.

Reply:
Thanks @Patol75 and @dsroberts I'm learning a lot about python and classes! 

_What if new functionalities must be defined before the methods are called but after the remainder of the parent's __init__ code?_

Do we have this scenario? Maybe it is more complicated in the update_approx branch with the stokes integrators, but cant you just define the extra attributes for the later clasess before you call the super().__init__ method? And then you could put the set_equations method etc just inside the init method of the abstract base class.

Reply:
Fair (provided it is compatible with an abstract base class; I have not checked). However, how is it simpler, more user-friendly, and more maintainable than the metaclass approach? If the user/developer forgets to add the decorator, then the solver will not work.

Reply:
@w3168 Yes, that would work here. However, it would not handle the case where `super` comes first. It makes sense to me that the parent's `__init__` should be called first, then the subclass's `__init__` adds to it, and finally `__post_init__` (or any other name for it) is called, having access to all `__init__` code.